Drop Python 2 Support

jhamman opened the pull request:
Drops support for Python 2. Closes #393.

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [x] Changes documented in docs/release.rst
* [x] Docs build locally (e.g., run ``tox -e docs``)
* [x] AppVeyor and Travis CI passes
* [x] Test coverage is 100% (Coveralls passes)

jeromekelleher commented:
Thanks for doing this @jhamman! We dropped Python 2 support in tskit in [this PR](https://github.com/tskit-dev/tskit/pull/140), which might be useful in outlining the steps required.  In particular, [pyupgrade](https://github.com/asottile/pyupgrade) seems really useful.

pep8speaks commented:
Hello @jhamman! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





























* In the file [`zarr/storage.py`](https://github.com/zarr-developers/zarr-python/blob/c9c52cd951ddfe6ce817105bce0ecfb50c4cbef5/zarr/storage.py):

> [Line 1526:1](https://github.com/zarr-developers/zarr-python/blob/c9c52cd951ddfe6ce817105bce0ecfb50c4cbef5/zarr/storage.py#L1526): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1























* In the file [`zarr/tests/test_storage.py`](https://github.com/zarr-developers/zarr-python/blob/c9c52cd951ddfe6ce817105bce0ecfb50c4cbef5/zarr/tests/test_storage.py):

> [Line 1018:1](https://github.com/zarr-developers/zarr-python/blob/c9c52cd951ddfe6ce817105bce0ecfb50c4cbef5/zarr/tests/test_storage.py#L1018): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

jhamman commented:
Thanks @jeromekelleher - I've implemented the following steps

1. changed setup.py to include `python_requires`
2. removed Python 2 special cases (e.g. `if PY2: ...`)
3. run `pyupgrade`
4. run `isort -rc --atomic .` (helped with step 5)
5. removed `zarr.compat`, imports to py3 modules

Tests are passing locally for me but we'll wait to see what coveralls/travis reports. I also need to add a note in `docs/release.rst`

jhamman commented:
@zarr-developers/core-devs - just pinging everyone on this. Luckily, we don't have any merge conflicts yet but it would still be good to get this reviewed and merged before things get messy.

jhamman commented:
@zarr-developers/core-devs - one more ping here. This needs a review.

alimanfoo commented on tox.ini:
```diff
@@ -4,15 +4,14 @@
 setenv =
     PYTHONHASHSEED = 42
     # hooks for coverage exclusions based on Python major version
     py35,py36,py37: PY_MAJOR_VERSION = py3
```
Actually this line could go as well, it's only needed for marking code to be excluded from coverage when running tests under py2, which we are not running any more.

jhamman replied:
Done. Thanks @alimanfo!

alimanfoo commented:
Thank you Joe, feels like a weight lifted to see this one go in :-)

On Sat, 19 Oct 2019, 05:10 Joe Hamman, <notifications@github.com> wrote:

> Merged #470 <https://github.com/zarr-developers/zarr-python/pull/470>
> into master.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zarr-developers/zarr-python/pull/470?email_source=notifications&email_token=AAFLYQRLXEO5I243FHYYVB3QPKCENA5CNFSM4ILMQDB2YY3PNVWWK3TUL52HS4DFWZEXG43VMVCXMZLOORHG65DJMZUWGYLUNFXW5KTDN5WW2ZLOORPWSZGOUKDYFTI#event-2726789837>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFLYQSKSV3YPZ2F6FBNSBTQPKCENANCNFSM4ILMQDBQ>
> .
>
